In [2]:
import pandas as pd
import numpy as np
from pandas import Series,DataFrame


In [9]:
import pymysql
conn = pymysql.connect(
    host='47.104.31.138',port=3306,
    user='guest',password='Guest.618',
    database='hrs', charset='utf8mb4'
    
)
dept_df = pd.read_sql('select * from tb_dept', conn, index_col='dno')
emp_df = pd.read_sql('select * from tb_emp', conn, index_col='eno')


#### 缺省值

In [12]:
    emp_df.isnull()

ename    job    mgr    sal   comm    dno
eno                                           
1359  False  False  False  False  False  False
2056  False  False  False  False  False  False
3088  False  False  False  False  False  False
3211  False  False  False  False   True  False
3233  False  False  False  False   True  False
3244  False  False  False  False   True  False
3251  False  False  False  False   True  False
3344  False  False  False  False  False  False
3577  False  False  False  False   True  False
3588  False  False  False  False   True  False
4466  False  False  False  False   True  False
5234  False  False  False  False   True  False
5566  False  False  False  False  False  False
7800  False  False   True  False  False  False

In [14]:
emp_df.isna()

ename    job    mgr    sal   comm    dno
eno                                           
1359  False  False  False  False  False  False
2056  False  False  False  False  False  False
3088  False  False  False  False  False  False
3211  False  False  False  False   True  False
3233  False  False  False  False   True  False
3244  False  False  False  False   True  False
3251  False  False  False  False   True  False
3344  False  False  False  False  False  False
3577  False  False  False  False   True  False
3588  False  False  False  False   True  False
4466  False  False  False  False   True  False
5234  False  False  False  False   True  False
5566  False  False  False  False  False  False
7800  False  False   True  False  False  False

相对应的，notnull和notna方法可以将非空的值标记为True。如果想删除这些缺失值，可以使用DataFrame对象的dropna方法，该方法的axis参数可以指定沿着0轴还是1轴删除，也就是说当遇到空值时，是删除整行还是删除整列，默认是沿0轴进行删除的，代码如下所示。

In [15]:
emp_df.dropna()

ename   job     mgr   sal    comm  dno
eno                                        
1359   胡一刀   销售员  3344.0  1800   200.0   30
2056    乔峰   分析师  7800.0  5000  1500.0   20
3088   李莫愁   设计师  2056.0  3500   800.0   20
3344    黄蓉  销售主管  7800.0  3000   800.0   30
5566   宋远桥   会计师  7800.0  4000  1000.0   10

In [17]:
#如果要沿着1轴进行删除，可以使用下面的代码。
emp_df.dropna(axis=1)

ename   job   sal  dno
eno                        
1359   胡一刀   销售员  1800   30
2056    乔峰   分析师  5000   20
3088   李莫愁   设计师  3500   20
3211   张无忌   程序员  3200   20
3233   丘处机   程序员  3400   20
3244   欧阳锋   程序员  3200   20
3251   张翠山   程序员  4000   20
3344    黄蓉  销售主管  3000   30
3577    杨过    会计  2200   10
3588   朱九真    会计  2500   10
4466   苗人凤   销售员  2500   30
5234    郭靖    出纳  2000   10
5566   宋远桥   会计师  4000   10
7800   张三丰    总裁  9000   20

注意：DataFrame对象的很多方法都有一个名为inplace的参数，该参数的默认值为False，表示我们的操作不会修改原来的DataFrame对象，而是将处理后的结果通过一个新的DataFrame对象返回。如果将该参数的值设置为True，那么我们的操作就会在原来的DataFrame上面直接修改，方法的返回值为None。简单的说，上面的操作并没有修改emp_df，而是返回了一个新的DataFrame对象。

在某些特定的场景下，我们可以对空值进行填充，对应的方法是fillna，填充空值时可以使用指定的值（通过value参数进行指定），也可以用表格中前一个单元格（通过设置参数method=ffill）或后一个单元格（通过设置参数method=bfill）的值进行填充，当代码如下所示。

In [20]:
emp_df.fillna(value=0)

ename   job     mgr   sal    comm  dno
eno                                        
1359   胡一刀   销售员  3344.0  1800   200.0   30
2056    乔峰   分析师  7800.0  5000  1500.0   20
3088   李莫愁   设计师  2056.0  3500   800.0   20
3211   张无忌   程序员  2056.0  3200     0.0   20
3233   丘处机   程序员  2056.0  3400     0.0   20
3244   欧阳锋   程序员  3088.0  3200     0.0   20
3251   张翠山   程序员  2056.0  4000     0.0   20
3344    黄蓉  销售主管  7800.0  3000   800.0   30
3577    杨过    会计  5566.0  2200     0.0   10
3588   朱九真    会计  5566.0  2500     0.0   10
4466   苗人凤   销售员  3344.0  2500     0.0   30
5234    郭靖    出纳  5566.0  2000     0.0   10
5566   宋远桥   会计师  7800.0  4000  1000.0   10
7800   张三丰    总裁     0.0  9000  1200.0   20

In [21]:
emp_df.fillna(method='backfill')

ename   job     mgr   sal    comm  dno
eno                                        
1359   胡一刀   销售员  3344.0  1800   200.0   30
2056    乔峰   分析师  7800.0  5000  1500.0   20
3088   李莫愁   设计师  2056.0  3500   800.0   20
3211   张无忌   程序员  2056.0  3200   800.0   20
3233   丘处机   程序员  2056.0  3400   800.0   20
3244   欧阳锋   程序员  3088.0  3200   800.0   20
3251   张翠山   程序员  2056.0  4000   800.0   20
3344    黄蓉  销售主管  7800.0  3000   800.0   30
3577    杨过    会计  5566.0  2200  1000.0   10
3588   朱九真    会计  5566.0  2500  1000.0   10
4466   苗人凤   销售员  3344.0  2500  1000.0   30
5234    郭靖    出纳  5566.0  2000  1000.0   10
5566   宋远桥   会计师  7800.0  4000  1000.0   10
7800   张三丰    总裁     NaN  9000  1200.0   20

In [23]:
emp_df.fillna(method='ffill')

ename   job     mgr   sal    comm  dno
eno                                        
1359   胡一刀   销售员  3344.0  1800   200.0   30
2056    乔峰   分析师  7800.0  5000  1500.0   20
3088   李莫愁   设计师  2056.0  3500   800.0   20
3211   张无忌   程序员  2056.0  3200   800.0   20
3233   丘处机   程序员  2056.0  3400   800.0   20
3244   欧阳锋   程序员  3088.0  3200   800.0   20
3251   张翠山   程序员  2056.0  4000   800.0   20
3344    黄蓉  销售主管  7800.0  3000   800.0   30
3577    杨过    会计  5566.0  2200   800.0   10
3588   朱九真    会计  5566.0  2500   800.0   10
4466   苗人凤   销售员  3344.0  2500   800.0   30
5234    郭靖    出纳  5566.0  2000   800.0   10
5566   宋远桥   会计师  7800.0  4000  1000.0   10
7800   张三丰    总裁  7800.0  9000  1200.0   20

注意：填充的值如何选择也是一个值得探讨的话题，实际工作中，可能会使用某种统计量（如：均值、众数等）进行填充，或者使用某种插值法（如：随机插值法、拉格朗日插值法等）进行填充，甚至有可能通过回归模型、贝叶斯模型等对缺失数据进行填充。b

#### 重复值

接下来，我们先给之前的部门表添加两行数据，让部门表中名为“研发部”和“销售部”的部门各有两个。

In [28]:
dept_df.loc[50] = {'dname': '研发部', 'dloc': '上海'}
dept_df.loc[60] = {'dname': '销售部', 'dloc': '长沙'}
dept_df

dname dloc
dno           
10    会计部   北京
20    研发部   成都
30    销售部   重庆
40    运维部   深圳
50    研发部   上海
60    销售部   长沙

In [30]:
dept_df

dname dloc
dno           
10    会计部   北京
20    研发部   成都
30    销售部   重庆
40    运维部   深圳
50    研发部   上海
60    销售部   长沙

In [29]:
dept_df.duplicated('dname')

dno
10    False
20    False
30    False
40    False
50     True
60     True
dtype: bool

从上面的输出可以看到，50和60两个部门从部门名称上来看是重复的，如果要删除重复值，可以使用drop_duplicates方法，该方法的keep参数可以控制在遇到重复值时，保留第一项还是保留最后一项，或者多个重复项一个都不用保留，全部删除掉。

In [31]:
dept_df.drop_duplicates('dname')

dname dloc
dno           
10    会计部   北京
20    研发部   成都
30    销售部   重庆
40    运维部   深圳

In [33]:
dept_df.drop_duplicates('dname', keep='last')

dname dloc
dno           
10    会计部   北京
40    运维部   深圳
50    研发部   上海
60    销售部   长沙

#### 异常值

异常值在统计学上的全称是疑似异常值，也称作离群点（outlier），异常值的分析也称作离群点分析。异常值是指样本中出现的“极端值”，数据值看起来异常大或异常小，其分布明显偏离其余的观测值。实际工作中，有些异常值可能是由系统或人为原因造成的，但有些异常值却不是，它们能够重复且稳定的出现，属于正常的极端值，例如很多游戏产品中头部玩家的数据往往都是离群的极端值。所以，我们既不能忽视异常值的存在，也不能简单地把异常值从数据分析中剔除。重视异常值的出现，分析其产生的原因，常常成为发现问题进而改进决策的契机。

异常值的检测有Z-score 方法、IQR 方法、DBScan 聚类、孤立森林等，这里我们对前两种方法做一个简单的介绍。

如果数据服从正态分布，依据3σ法则，异常值被定义与平均值的偏差超过三倍标准差的值。在正态分布下，距离平均值3σ之外的值出现的概率为$ P(|x-\mu|>3\sigma)<0.003 $，属于小概率事件。如果数据不服从正态分布，那么可以用远离平均值的多少倍的标准差来描述，这里的倍数就是Z-score。Z-score以标准差为单位去度量某一原始分数偏离平均值的距离，公式如下所示。 $$ z = \frac {X - \mu} {\sigma} $$ Z-score需要根据经验和实际情况来决定，通常把远离标准差3倍距离以上的数据点视为离群点，下面的代给出了如何通过Z-score方法检测异常值。

In [39]:
def detect_outliers_zscore(data, threshold=3):
    avg_value = np.mean(data)# 平均值
    std_value = np.std(data)# 标准差
    z_score = np.abs((data - avg_value) / std_value)
    return data[z_score > threshold]


IQR 方法中的IQR（Inter-Quartile Range）代表四分位距离，即上四分位数（Q3）和下四分位数（Q1）的差值。通常情况下，可以认为小于 $ Q1 - 1.5 \times IQR $ 或大于 $ Q3 + 1.5 \times IQR $ 的就是m异常值，而这种检测异常值的方法也是箱线图（后面会讲到）默认使用的方法。下面的代给出了如何通过 IQR 方法检测异常值。

In [47]:
import numpy as np
def detect_outliers_iqr(data, whis=1.5):
    q1, q3 = np.quantile(data, [0.25, 0.75])
    iqr = q3 - q1
    lower, upper = q1 - whis * iqr, q3 + whis * iqr
    return data[(data < lower) | (data > upper)]

如果要删除异常值，可以使用DataFrame对象的drop方法，该方法可以根据行索引或列索引删除指定的行或列。例如我们认为月薪低于2000或高于8000的是员工表中的异常值，可以用下面的代码删除对应的记录。

In [52]:
emp_df.drop(emp_df[(emp_df.sal>8000)|(emp_df.sal<2000)].index)

ename   job     mgr   sal    comm  dno
eno                                        
2056    乔峰   分析师  7800.0  5000  1500.0   20
3088   李莫愁   设计师  2056.0  3500   800.0   20
3211   张无忌   程序员  2056.0  3200     NaN   20
3233   丘处机   程序员  2056.0  3400     NaN   20
3244   欧阳锋   程序员  3088.0  3200     NaN   20
3251   张翠山   程序员  2056.0  4000     NaN   20
3344    黄蓉  销售主管  7800.0  3000   800.0   30
3577    杨过    会计  5566.0  2200     NaN   10
3588   朱九真    会计  5566.0  2500     NaN   10
4466   苗人凤   销售员  3344.0  2500     NaN   30
5234    郭靖    出纳  5566.0  2000     NaN   10
5566   宋远桥   会计师  7800.0  4000  1000.0   10

如果要替换掉异常值，可以通过给单元格赋值的方式来实现，也可以使用replace方法将指定的值替换掉。例如我们要将月薪为1800和9000的替换为月薪的平均值，补贴为800的替换为1000，代码如下所示。

In [64]:
avg_sal = np.mean(emp_df.sal)
emp_df.replace({'sal': [1800, 9000], 'comm': 800}, {'sal': avg_sal, 'comm': 1000})

ename   job     mgr          sal    comm  dno
eno                                               
1359   胡一刀   销售员  3344.0  3521.428571   200.0   30
2056    乔峰   分析师  7800.0  5000.000000  1500.0   20
3088   李莫愁   设计师  2056.0  3500.000000  1000.0   20
3211   张无忌   程序员  2056.0  3200.000000     NaN   20
3233   丘处机   程序员  2056.0  3400.000000     NaN   20
3244   欧阳锋   程序员  3088.0  3200.000000     NaN   20
3251   张翠山   程序员  2056.0  4000.000000     NaN   20
3344    黄蓉  销售主管  7800.0  3000.000000  1000.0   30
3577    杨过    会计  5566.0  2200.000000     NaN   10
3588   朱九真    会计  5566.0  2500.000000     NaN   10
4466   苗人凤   销售员  3344.0  2500.000000     NaN   30
5234    郭靖    出纳  5566.0  2000.000000     NaN   10
5566   宋远桥   会计师  7800.0  4000.000000  1000.0   10
7800   张三丰    总裁     NaN  3521.428571  1200.0   20

#### 预处理

对数据进行预处理也是一个很大的话题，它包含了对数据的拆解、变换、归约、离散化等操作。我们先来看看数据的拆解。如果数据表中的数据是一个时间日期，我们通常都需要从年、季度、月、日、星期、小时、分钟等维度对其进行拆解，如果时间日期是用字符串表示的，可以先通过pandas的to_datetime函数将其处理成时间日期。

在下面的例子中，我们先读取 Excel 文件，获取到一组销售数据，其中第一列就是销售日期，我们将其拆解为“月份”、“季度”和“星期”，代码如下所示。

In [70]:
sales_df = pd.read_excel(
    'res/2020年销售数据.xlsx',
    usecols = ['销售日期', '销售区域', '销售渠道', '品牌', '销售额']
)
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1945 entries, 0 to 1944
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   销售日期    1945 non-null   datetime64[ns]
 1   销售区域    1945 non-null   object        
 2   销售渠道    1945 non-null   object        
 3   品牌      1945 non-null   object        
 4   销售额     1945 non-null   int64         
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 76.1+ KB


In [73]:
sales_df

销售日期 销售区域 销售渠道    品牌    销售额
0    2020-01-01   上海  拼多多   八匹马   8217
1    2020-01-01   上海   抖音   八匹马   6351
2    2020-01-01   上海   天猫   八匹马  14365
3    2020-01-01   上海   天猫   八匹马   2366
4    2020-01-01   上海   天猫   皮皮虾  15189
...         ...  ...  ...   ...    ...
1940 2020-12-30   北京   京东  花花姑娘   6994
1941 2020-12-30   福建   实体   八匹马   7663
1942 2020-12-31   福建   实体  花花姑娘  14795
1943 2020-12-31   福建   抖音   八匹马   3481
1944 2020-12-31   福建   天猫   八匹马   2673

[1945 rows x 5 columns]

In [76]:
sales_df['月份']=sales_df['销售日期'].dt.month
sales_df['季度']=sales_df['销售日期'].dt.quarter
sales_df['星期']=sales_df['销售日期'].dt.weekday
sales_df

销售日期 销售区域 销售渠道    品牌    销售额  月份  季度  星期
0    2020-01-01   上海  拼多多   八匹马   8217   1   1   2
1    2020-01-01   上海   抖音   八匹马   6351   1   1   2
2    2020-01-01   上海   天猫   八匹马  14365   1   1   2
3    2020-01-01   上海   天猫   八匹马   2366   1   1   2
4    2020-01-01   上海   天猫   皮皮虾  15189   1   1   2
...         ...  ...  ...   ...    ...  ..  ..  ..
1940 2020-12-30   北京   京东  花花姑娘   6994  12   4   2
1941 2020-12-30   福建   实体   八匹马   7663  12   4   2
1942 2020-12-31   福建   实体  花花姑娘  14795  12   4   3
1943 2020-12-31   福建   抖音   八匹马   3481  12   4   3
1944 2020-12-31   福建   天猫   八匹马   2673  12   4   3

[1945 rows x 8 columns]

在上面的代码中，通过日期时间类型的Series对象的dt 属性，获得一个访问日期时间的对象，通过该对象的year、month、quarter、hour等属性，就可以获取到年、月、季度、小时等时间信息，获取到的仍然是一个Series对象，它包含了一组时间信息，所以我们通常也将这个dt属性称为“日期时间向量”。

我们再来说一说字符串类型的数据的处理，我们先从指定的 Excel 文件中读取某招聘网站的招聘数据。

In [79]:
job_df = pd.read_csv('res/某招聘网站招聘数据.csv',
                    usecols=['city', 'companyFullName', 'positionName', 'salary'])
job_df

city   companyFullName  positionName   salary
0      北京    达疆网络科技（上海）有限公司         数据分析岗  15k-30k
1      北京      北京音娱时光科技有限公司          数据分析  10k-18k
2      北京     北京千喜鹤餐饮管理有限公司          数据分析  20k-30k
3      北京     吉林省海生电子商务有限公司          数据分析  33k-50k
4      北京    韦博网讯科技（北京）有限公司          数据分析  10k-15k
...   ...               ...           ...      ...
3135   天津  清博津商（天津）教育科技有限公司   审核实习生（春节短期）    1k-2k
3136   天津  上海礼紫股权投资基金管理有限公司          运营助理    6k-8k
3137   天津    北京达佳互联信息技术有限公司  运营编辑团队leader   8k-15k
3138   天津     北京河狸家信息技术有限公司       商家运营-天津    6k-8k
3139   天津     北京河狸家信息技术有限公司      运营实习生-天津    2k-4k

[3140 rows x 4 columns]

In [80]:
job_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3140 entries, 0 to 3139
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   city             3140 non-null   object
 1   companyFullName  3140 non-null   object
 2   positionName     3140 non-null   object
 3   salary           3140 non-null   object
dtypes: object(4)
memory usage: 98.2+ KB


In [83]:
job_df.head()

city companyFullName positionName   salary
0   北京  达疆网络科技（上海）有限公司        数据分析岗  15k-30k
1   北京    北京音娱时光科技有限公司         数据分析  10k-18k
2   北京   北京千喜鹤餐饮管理有限公司         数据分析  20k-30k
3   北京   吉林省海生电子商务有限公司         数据分析  33k-50k
4   北京  韦博网讯科技（北京）有限公司         数据分析  10k-15k

上面的数据表一共有3140条数据，但并非所有的职位都是“数据分析”的岗位，如果要筛选出数据分析的岗位，可以通过检查positionName字段是否包含“数据分析”这个关键词，这里需要模糊匹配，应该如何实现呢？我们可以先获取positionName列，因为这个Series对象的dtype是字符串，所以可以通过str属性获取对应的字符串向量，然后就可以利用我们熟悉的字符串的方法来对其进行操作，代码如下所示。

In [100]:
job_df.positionName.dtype

dtype('O')

In [88]:
np.dtype(job_df.positionName)

dtype('O')

In [110]:
job_df.positionName.str.contains('数据分析')

0       True
1       True
2       True
3       True
4       True
        ... 
3065    True
3069    True
3070    True
3071    True
3088    True
Name: positionName, Length: 1515, dtype: bool

In [103]:
#用来描述与数组对应的内存区域是如何使用，它描述了数据的以下几个方
job_df = job_df[job_df.positionName.str.contains('数据分析')]

In [112]:
job_df.shape

(1515, 4)

In [113]:
job_df


city   companyFullName     positionName   salary
0      北京    达疆网络科技（上海）有限公司            数据分析岗  15k-30k
1      北京      北京音娱时光科技有限公司             数据分析  10k-18k
2      北京     北京千喜鹤餐饮管理有限公司             数据分析  20k-30k
3      北京     吉林省海生电子商务有限公司             数据分析  33k-50k
4      北京    韦博网讯科技（北京）有限公司             数据分析  10k-15k
...   ...               ...              ...      ...
3065   天津    北京达佳互联信息技术有限公司           数据分析专员   8k-10k
3069   天津  林之尚沃（天津）国际贸易有限公司           数据分析专员   6k-10k
3070   天津  林之尚沃（天津）国际贸易有限公司          数据分析实习生    2k-4k
3071   天津        西安格蒂电力有限公司  ETL/大数据/数据分析/实施   6k-12k
3088   天津          谷川联行有限公司  BI工程师/数据分析师（天津）   8k-12k

[1515 rows x 4 columns]

可以看出，筛选后的数据还有1515条。接下来，我们还需要对salary字段进行处理，如果我们希望统计所有岗位的平均工资或每个城市的平均工资，首先需要将用范围表示的工资处理成其中间值，代码如下所示。

In [115]:
job_df.salary.str.extract(r'(\d+)[kK]?-(\d+)[kK]?')

0   1
0     15  30
1     10  18
2     20  30
3     33  50
4     10  15
...   ..  ..
3065   8  10
3069   6  10
3070   2   4
3071   6  12
3088   8  12

[1515 rows x 2 columns]

需要提醒大家的是，抽取出来的两列数据都是字符串类型的值，我们需要将其转换成int类型，才能计算平均值，对应的方法是DataFrame对象的applymap方法，该方法的参数是一个函数，而该函数会作用于DataFrame中的每个元素。完成这一步之后，我们就可以使用apply方法将上面的DataFrame处理成中间值，apply方法的参数也是一个函数，可以通过指定axis参数使其作用于DataFrame 对象的行或列，代码如下所

In [125]:
job_df.salary.str.extract(r'(\d+)[kK]?-(\d+)[kK]?').applymap(int)

0   1
0     15  30
1     10  18
2     20  30
3     33  50
4     10  15
...   ..  ..
3065   8  10
3069   6  10
3070   2   4
3071   6  12
3088   8  12

[1515 rows x 2 columns]

In [130]:
temp_df = job_df.salary.str.extract(r'(\d+)[kK]?-(\d+)[kK]?').applymap(int)#str转换成int
#apply方法的参数也是一个函数，可以通过指定axis参数使其作用于DataFrame 对象的行或列，代码如下所
temp_df.apply(np.mean, axis=1)

0       22.5
1       14.0
2       25.0
3       41.5
4       12.5
        ... 
3065     9.0
3069     8.0
3070     3.0
3071     9.0
3088    10.0
Length: 1515, dtype: float64

接下来，我们可以用上面的结果替换掉原来的salary列或者增加一个新的列来表示职位对应的工资，完整的代码如下所示。

In [ ]:
temp_df = jobs_df.salary.str.extract(r'(\d+)[kK]?-(\d+)[kK]?').applymap(int)
jobs_df['salary'] = temp_df.apply(np.mean, axis=1)
jobs_df.head()

applymap和apply两个方法在数据预处理的时候经常用到，Series对象也有apply方法，也是用于数据的预处理，但是DataFrame对象还有一个名为transform 的方法，也是通过传入的函数对数据进行变换，类似Series对象的map方法。需要强调的是，apply方法具有归约效果的，简单的说就是能将较多的数据处理成较少的数据或一条数据；而transform方法没有归约效果，只能对数据进行变换，原来有多少条数据，处理后还是有多少条数据。

如果要对数据进行深度的分析和挖掘，字符串、日期时间这样的非数值类型都需要处理成数值，因为非数值类型没有办法计算相关性，也没有办法进行$\chi^2$检验等操作。对于字符串类型，通常可以其分为以下三类，再进行对应的处理。

1、有序变量（Ordinal Variable）：字符串表示的数据有顺序关系，那么可以对字符串进行序号化处理。

2、分类变量（Categorical Variable）/ 名义变量（Nominal Variable）：字符串表示的数据没有大小关系和等级之分，那么就可以使用独热编码的方式处理成哑变量（虚拟变量）矩阵。

3、定距变量（Scale Variable）：字符串本质上对应到一个有大小高低之分的数据，而且可以进行加减运算，那么只需要将字符串处理成对应的数值即可

对于第1类和第3类，我们可以用上面提到的apply或transform方法来处理，也可以利用scikit-learn中的OrdinalEncoder处理第1类字符串，这个我们在后续的课程中会讲到。对于第2类字符串，可以使用pandas的get_dummies()函数来生成哑变量（虚拟变量）矩阵，代码如下所示。

In [138]:
persons_df = pd.DataFrame(
        data={
            '姓名':['关羽', '张飞', '赵云', '马超', '黄忠'], 
            '职业':['医生', '医生', '程序员', '画家', '教师'],
            '学历':['研究生', '大专', '研究生', '高中', '本科']
        }
)
persons_df

姓名   职业   学历
0  关羽   医生  研究生
1  张飞   医生   大专
2  赵云  程序员  研究生
3  马超   画家   高中
4  黄忠   教师   本科

将职业处理成哑变量矩阵。

In [140]:
pd.get_dummies(persons_df['职业'])

医生  教师  画家  程序员
0   1   0   0    0
1   1   0   0    0
2   0   0   0    1
3   0   0   1    0
4   0   1   0    0

将学历处理成大小不同的值。

In [142]:
persons_df['学历']

0    研究生
1     大专
2    研究生
3     高中
4     本科
Name: 学历, dtype: object

In [143]:
def handle_education(x):
    edu_dict = {'高中': 1, '大专': 3, '本科': 5, '研究生': 10}
    return edu_dict.get(x, 0)
#apply方法的参数也是一个函数，可以通过指定axis参数使其作用于DataFrame 对象的行或列，代码如下所,作用于每一个元素
persons_df['学历'].apply(handle_education)

0    10
1     3
2    10
3     1
4     5
Name: 学历, dtype: int64

我们再来说说数据离散化。离散化也叫分箱，如果变量的取值是连续值，那么它的取值有无数种可能，在进行数据分组的时候就会非常的不方便，这个时候将连续变量离散化就显得非常重要。之所以把离散化叫做分箱，是因为我们可以预先设置一些箱子，每个箱子代表了数据取值的范围，这样就可以将连续的值分配到不同的箱子中，从而实现离散化。下面的例子读取了2018年北京积分落户数据，我们可以根据落户积分对数据进行分组，具体的做法如下所示。

In [174]:
luohu_df = pd.read_csv('res/2018年北京积分落户数据.csv', index_col='id')
luohu_df.score.describe()

count    6019.000000
mean       95.654552
std         4.354445
min        90.750000
25%        92.330000
50%        94.460000
75%        97.750000
max       122.590000
Name: score, dtype: float64

可以看出，落户积分的最大值是122.59，最小值是90.75，那么我们可以构造一个从90分到125分，每5分一组的7个箱子，pandas的cut函数可以帮助我们首先数据分箱，代码如下所示。

In [172]:
list('ABCDEFG')


['A', 'B', 'C', 'D', 'E', 'F', 'G']

In [189]:
bins = np.arange(90,126,7)
bins
pd_cut = pd.cut(luohu_df.score, bins ,labels = list('ABCDE') ,right = False)
pd_cut

id
1       E
2       E
3       E
4       E
5       D
       ..
6015    A
6016    A
6017    A
6018    A
6019    A
Name: score, Length: 6019, dtype: category
Categories (5, object): [A < B < C < D < E]

说明：cut函数的right参数默认值为True，表示箱子左开右闭；修改为False可以让箱子的右边界为开区间，左边界为闭区间，大家看看上面的输出就铭板了。
我们可以根据分箱的结果对数据进行分组，然后使用聚合函数对每个组进行统计，这是数据分析中经常用到的操作，下一个章节会为大家介绍。除此之外，pandas还提供了一个名为qcut的函数，可以指定分位数对数据进行分箱，有兴趣的读者可以自行研究。


In [ ]:
 count = pd_cut.value_counts()
# pd_cut.plot(kind = "bar")
count.plot(kind = "bar")

In [190]:
pd_cut.index


Int64Index([   1,    2,    3,    4,    5,    6,    7,    8,    9,   10,
            ...
            6010, 6011, 6012, 6013, 6014, 6015, 6016, 6017, 6018, 6019],
           dtype='int64', name='id', length=6019)